### DISCUSSION TITLE
Copying signature file onto a USB stick for imgverify.

### DISCUSSION THREAD

**Discussion by adroutley** [ipxe/ipxe#1300] [Q&A] [answered]
Dear ipxe uses,

after i network-download a kernel, I check it with imgverify against the signature as per the documentation:-

https://ipxe.org/cmd/imgverify

Downloading the signature file via HTTP works like a charm, but I prefer to store it locally on the USB stick itself, as per the example in the documentation:

imgverify script.ipxe script.ipxe.sig

how would I get the signature file "script.ipxe.sig" onto the USB stick? I have tried the "make EMBED=" directive (like I do when I copy a script) but imgverify doesn't find it.

this ancient blog describes the problem perfectly:-
https://forum.ipxe.org/showthread.php?tid=31481

Any help with examples would be appreciated.

PART 2 OF THE SAME QUESTION:- 
how do I list the content of an IPXE usb image ?  I can't seem to mount it so I don't know what's in it....

Cheers!

**Comment by NiKiZe** [comment]
For efi it is a fat filesystem that you can access thru `file://`

For pcbios it is the bootloader only, no filesystem, and also there is no filesystem access. I do think embed can handle multiple files you could test with `EMBED=script,script.sig` and then check what you have with imgstat.

**Reply by adroutley** [reply]
hi @NiKiZe 

thanks for your sharp eyes.

per your suggestion, i changed the make syntax to this:-

make DEBUG=tls bin/ipxe.usb CERT=/home/andrew/bin/myipxe/openssl/ca.crt TRUST=/home/andrew/bin/myipxe/openssl/ca.crt EMBED=/home/andrew/bin/myipxe/netboot-client-SIG.ipxe,/home/andrew/bin/myipxe/openssl/vmlinuz-6.0.10-221201-nvidia-SIG,/home/andrew/bin/myipxe/openssl/initrd-6.0.10-221201-nvidia-SIG

...and now all the downloaded files and the embedded signatures appear in imgstat:-

http://www.andrewroutley.net/internet/ipxe/itworks.jpg

it works.  good catch-thanks...

**Comment by adroutley** [comment]
Thanks for your reply christian.

per your suggestion i loaded the -SIG file onto the USB stick via the EMBED
command:-

make DEBUG=tls bin/ipxe.usb CERT=/home/andrew/bin/myipxe/openssl/ca.crt
TRUST=/home/andrew/bin/myipxe/openssl/ca.crt
EMBED=/home/andrew/bin/myipxe/openssl/vmlinuz-6.0.10-221201-nvidia-SIG
EMBED=/home/andrew/bin/myipxe/openssl/initrd-6.0.10-221201-nvidia-SIG
EMBED=/home/andrew/bin/myipxe/netboot-client-SIG.ipxe

during the boot, the kernel and initrd download via http, and show up in
imgstat:
http://www.andrewroutley.net/internet/ipxe/imgstat.jpg

....but not the -SIG file that i loaded on to the USB stick directly.  it
then
hangs here:-

imgverify initrd-6.0.10-221201-nvidia initrd-6.0.10-221201-nvidia-SIG

 the full script is here:
http://www.andrewroutley.net/internet/ipxe/netboot-client-SIG.ipxe

any thoughts on why my EMBED commands are not working?

has anyone copied a signature to the USB stick?

cheers

**Reply by NiKiZe** [reply]
What you did there was to override the EMBED setting with each instance, try EMBED=script file.ipxe,otherfile.sig,otherfile2.sig

**Comment by mcb30** [comment, marked as answer]
The short answer is that you need to use a comma-separated `EMBED=script.ipxe,vmlinuz.sig,initrd.sig,...` (as stated above by @NiKiZe).

The slightly longer answer is: don't do this.  The whole point of using CMS signatures with X.509 certificates (rather than a simpler scheme such as just comparing the SHA-256 checksum of the expected file) is that you **don't** need to update anything on the client when you create a new signed image.  The signature file contains the SHA-256 checksum, and the certificate trusted via `TRUST=...` provides the ability to verify this checksum.  The certificate (embedded in the binary) doesn't need to change, even when you create a new version of the kernel (and hence a new signature).

You gain **zero** security by embedding the signatures into your iPXE binary.  You just make your own life more inconvenient, since you now have to update your iPXE binary whenever you update your kernel.

**Reply by adroutley** [reply]
hi @mcb30 

thanks for your feedback.  it is appreciated.

i got the whole idea and the syntax from the ipxe documentation:

https://ipxe.org/crypto#code_signing

as you can see, they allow either a network or a USB signature to verify the kernel.

the reasoning behind placing the signature on the stick is, a man-in-the-middle attack can theoretically compromise the kernel, but not the signature if it is "hidden" on the stick.  if both are loaded from the network, both are vulnerable, no?  (feel free to challenge this logic, as i am always open to suggestions.  ;)   )

as for "CMS signatures with X.509 certificates" in your feedback, would you please send me a blog with step by step instructions on how to set this up?  the more detail the better. the only thing i am familiar with is what i found in the docs.

again, thx 4 the feedback...

**Reply by adroutley** [reply]
whoops @mcb30 i should have sent this regarding USB signatures:

https://ipxe.org/cmd/imgverify

imgverify script.ipxe script.ipxe.sig

**Reply by NiKiZe** [reply]
As mcb says, don't do this. Verify the .sig files by certificate instead. If not you might as well keep kernel, initrd etc on the stick and avoid the potential mitm on network

**Reply by mcb30** [reply]
> i got the whole idea and the syntax from the ipxe documentation:
> 
> https://ipxe.org/crypto#code_signing
> ...
> as for "CMS signatures with X.509 certificates" in your feedback, would you please send me a blog with step by step instructions on how to set this up? the more detail the better. the only thing i am familiar with is what i found in the docs.

If you're following that guide and using `imgverify`, then you're already using CMS signatures with X.509 certificates.

> the reasoning behind placing the signature on the stick is, a man-in-the-middle attack can theoretically compromise the kernel, but not the signature if it is "hidden" on the stick. if both are loaded from the network, both are vulnerable, no? (feel free to challenge this logic, as i am always open to suggestions. ;) )

No, the signature is not vulnerable even if sent over an untrusted network (and subject to a MITM attack).  The signature is itself signed by the certificate that iPXE trusts (`/home/andrew/bin/myipxe/openssl/ca.crt` in your example).  Without the private key corresponding to this certificate, there is no way for an attacker to forge a signature that will be accepted by iPXE.

**Reply by adroutley** [reply]
"The signature is itself signed by the certificate that iPXE trusts"

okay this was the missing part of the puzzle.  I was aware of course that i created a kernel signature:-

openssl cms -sign -binary -noattr -in vmlinuz-6.0.10-221201-nvidia -signer codesign.crt -inkey codesign.key -certfile ca.crt -outform DER -out vmlinuz-6.0.10-221201-nvidia-SIG

...but I did not know that the signature, itself, is also signed, and checked by IPXE (and therefore not vulnerable to any MITM attack).  Network loading is therefore the best route as you and @NiKiZe  suggest.

So....if I have a new kernel I just create a new signature and place them both on the server ?

**Reply by mcb30** [reply]
> So....if I have a new kernel I just create a new signature and place them both on the server ?

Yes, exactly that.
